### ISSUE TITLE
Missing avx2-i64x8 target (ie. double pumping for 64-bit codes)

### ISSUE THREAD

**Issue by zephyr111** [ispc/ispc#2903]
Hello,

There is no `avx2-i64x8` target in ISPC while there is a `avx2-i32x16` target to enable double-pumping on 32-bit codes.
While ISPC is mainly design to speed up 32-bit codes, it can be very helpful to speed up 64-bit ones. Providing this missing target can help to significantly reduce the execution time of some 64-bit code on machines supporting AVX-2 but not AVX-512 yet (i.e. ~80% machines), especially since the latency of 64-bit instructions tends to be higher than 32-bit ones. 

Note it is not a problem with AVX-512 anymore since we can use targets like avx512skx-x8 or avx512skx-x16, but only a tiny fraction of mainstream client (Intel) CPUs support it (i.e. <15%), and even the newest client Intel CPUs still don't support it (so generating AVX-2 code will still be useful for >5 years).

Here is a simple example:

```
uniform double sum(uniform double v[], uniform size_t size)
{
    varying double s = 0.0d;
    foreach(i=0...size)
        s += v[i];
    return reduce_add(s);
}
```

Here is the generated code for the `avx2-i64x4` target ([see on GodBolt](https://godbolt.org/z/64r7f6c1j)):

```
.LBB1_3:
        mov     r8d, edx
        and     r8d, -32
        vaddpd  ymm1, ymm1, ymmword ptr [rdi + r8]
        add     rax, 4
        add     rdx, 32
        cmp     rax, rcx
        jb      .LBB1_3
```

This takes 4 cycles/iteration on all Intel CPU from Skylake to Rocket-Lake. 3 cycles/iteration on (P-cores of) newer Intel CPUs and all AMD Zen ones.

With double-pumping (ie. a new `avx2-i64x8` target), the following generated code is expected:

```
.LBB1_3:                                # =>This Inner Loop Header: Depth=1
        mov     r8d, edx
        and     r8d, -64
        vaddpd  ymm2, ymm2, ymmword ptr [rdi + r8 + 32]
        vaddpd  ymm1, ymm1, ymmword ptr [rdi + r8]
        add     rax, 8
        add     rdx, 64
        cmp     rax, rcx
        jb      .LBB1_3
```

It should take exactly the same number of cycles/iteration while computing 8 items/iteration rather than 4 resulting in a **twice faster code**. The code is still not optimal but it should be enough in most case to mitigate the latency in more complex use-cases.

Other examples where this can be *very-useful* include: computing functions like `sin`/`cos`/`exp`/`log` in a SIMD-friendly way and more generally all high-order polynomials (approximations), Kahan summation.

Can this target be supported?

**Comment by aneshlya** [comment]
I see the code you expect with `avx2-i32x8` target. 
Let's clarify to be on the same page that `i32`/`i64` etc in ISPC target name stands for mask size, not an addressing calculation (you can use `--addressing` option for this).
Would `avx2-i32x8` work for your case?

**Comment by zephyr111** [comment]
This is the "cheat" I currently used so far so to improve performance of most double-precision codes (and the one I used to generate the above example), but sometimes it generates less efficient instructions than with `avx2-i64x4`. Still double-pumping help but its benefit can be significantly reduced by this problem. Besides, I did not know if this "cheat" was really a clean solution.

Here is one simple example which is a variation of the previous code so to discard NaN values:
```
uniform double sum(uniform double v[], uniform size_t size)
{
    varying double s = 0.0d;
    foreach(i=0...size)
        s += select(!isnan(v[i]), v[i], 0.0d);
    return reduce_add(s);
}
```
Here is the generated code for the avx2-i64x4 target (and 64-bit addressing, [see on GodBolt](https://godbolt.org/z/MhYYcojbE)):
```
.LBB1_3:                                # =>This Inner Loop Header: Depth=1
        vmovupd ymm3, ymmword ptr [rdi + 8*rax]
        vcmpordpd       ymm4, ymm3, ymm2
        vandpd  ymm3, ymm4, ymm3
        vaddpd  ymm1, ymm1, ymm3
        add     rax, 4
        cmp     rax, rcx
        jb      .LBB1_3
```
With double-pumping (ie. a new avx2-i64x8 target), the following generated code is expected:
```
.LBB1_3:                                # =>This Inner Loop Header: Depth=1
        vmovupd ymm4, ymmword ptr [rdi + 8*rax]
        vmovupd ymm5, ymmword ptr [rdi + 8*rax + 32]
        vcmpordpd       ymm6, ymm5, ymm3
        vcmpordpd       ymm7, ymm4, ymm3
        vandpd  ymm4, ymm6, ymm4
        vandpd  ymm5, ymm7, ymm5
        vaddpd  ymm1, ymm1, ymm4
        vaddpd  ymm2, ymm2, ymm5
        add     rax, 8
        cmp     rax, rcx
        jb      .LBB1_3
```
But, here is the generated code for the avx2-i32x8 target (and 64-bit addressing, [see on GodBolt](https://godbolt.org/z/b9xKvPjW7)):
```
.LBB1_3:                                # =>This Inner Loop Header: Depth=1
        vmovupd ymm4, ymmword ptr [rdi + 8*rax]
        vmovupd ymm5, ymmword ptr [rdi + 8*rax + 32]
        vcmpordpd       ymm6, ymm5, ymm3
        vcmpordpd       ymm7, ymm4, ymm3
        vpackssdw       ymm6, ymm7, ymm6
        vpshufd ymm7, ymm6, 80                  # ymm7 = ymm6[0,0,1,1,4,4,5,5]
        vblendvps       ymm4, ymm3, ymm4, ymm7
        vpshufd ymm6, ymm6, 250                 # ymm6 = ymm6[2,2,3,3,6,6,7,7]
        vblendvps       ymm5, ymm3, ymm5, ymm6
        vaddpd  ymm1, ymm1, ymm4
        vaddpd  ymm2, ymm2, ymm5
        add     rax, 8
        cmp     rax, rcx
        jb      .LBB1_3
```

There are unnecessary instructions (which seems to be generate due to the reduced 32-bit mask size not fitting with the 64-bit item size) which can make more-complex real-world code slower (typically when masking is performed). Indeed, the expected generated code typically results in 10 uops while the current generated typically results in 15 uops. This is 50% more uops!

In this specific use-case, on Skylake and Icelake, the dependency of the SIMD accumulator and the relatively-high latency of `vaddpd` hide the problem so the performance is not (much) affected when run sequentially. In parallel, with SMT enabled, the later code is slower because (shared) ports are saturated and there are more uops to execute/issue. On newer architectures like Alderlake and Sapphire Rapids, the latency of `vaddpd` is twice smaller (only 2 cycles) so I expect the later code to be slower too. Also please note that 32-bit addressing add more issues making this problem more visible in practice.

Here is another slightly-more complex example where this is slower on my Skylake machine (25-30% with 64-bit addressing and a bit more in 32-bit one) because of back-end port saturation: 

```
uniform double mean(uniform double v[], uniform size_t size)
{
    varying double s = 0.0d;
    varying int64 count = 0;
    foreach(i=0...size)
    {
        s += select(!isnan(v[i]), v[i], 0.0d);
        count += isnan(v[i]);
    }
    return reduce_add(s) / reduce_add(count);
}
```

**Comment by aneshlya** [comment]
I'm not sure that a new target will help here. @nurmukhametov, you were dealing with quite a lot of code generation problems recently. Do you have any thoughts on this matter?

**Comment by nurmukhametov** [comment]
> Here is a simple example:
> 
> ```
> uniform double sum(uniform double v[], uniform size_t size)
> {
>     varying double s = 0.0d;
>     foreach(i=0...size)
>         s += v[i];
>     return reduce_add(s);
> }
> ```
> 

I would try `#pragma unroll` for this example, e.g., ISPC generates with `#pragram unroll 4` the following code:

```asm
.LBB1_5:                                # =>This Inner Loop Header: Depth=1
        mov     r9d, r8d
        and     r9d, -128
        vaddpd  ymm1, ymm1, ymmword ptr [rdi + r9]
        vaddpd  ymm1, ymm1, ymmword ptr [rdi + r9 + 32]
        vaddpd  ymm1, ymm1, ymmword ptr [rdi + r9 + 64]
        vaddpd  ymm1, ymm1, ymmword ptr [rdi + r9 + 96]
        add     rax, 16
        sub     r8, -128
        add     edx, -4
        jne     .LBB1_5
```

**Comment by zephyr111** [comment]
`#pragma unroll` generates a pretty inefficient code here because of the dependency on `ymm1`. Indeed, the number of cycles of **the dependency chain is left unchanged with `#pragma unroll` so it does not improve performance in practice** (as long as instruction latency is the bottleneck which is the case on my Coffee-Lake CPU and most mainstream CPUs). To be more precise, on Skylake-like CPUs, the 4 `vaddpd` takes at least 16 cycles for computing 4 items while the initial loop takes 4 cycles for computing 1 item. Please note the presence of the `ymm2` register in the above `avx2-i64x8` example. This is the key to solve this performance issue and this is why double-pumping is so useful here.

By the way, the compiler cannot reorder dependent FP instructions because of FP non-associativity of FP numbers defined in the IEEE-754 standard. Thus, I expected this dependency chain to be released by the use of `--opt=fast-math` like with `-ffast-math` but it is surprisingly not. But I think this is another independent issue. Since fast-math is sometime dangerous (for example it can break the above example skipping NaN values since NaN values should not exists with fast-math) I think it is preferable not to rely on this to solve such a problem.

**Comment by pbrubaker** [comment]
Won't register renaming eliminate part of the dependency on `ymm1`?

**Comment by nurmukhametov** [comment]
> Thus, I expected this dependency chain issue to be fixed by the use of --opt=fast-math (like with -ffast-math in Clang), but it is surprisingly not the case with ISPC.

It is not the case with clang either, see https://godbolt.org/z/fh4q8zc8G

**Comment by zephyr111** [comment]
> Won't register renaming eliminate part of the dependency on `ymm1`?

Not in this case. Register renaming would help if it was a false dependency, but for true ones (like here), the processor must execute the `vaddpd` instructions in-order so to guarantee the strict IEEE-754 compliance (there are many cases where this is critical like in the Kahan summation, or if data are sorted in a way to reduce the final error for example). To be more precise, note the uops related to the actual addition is the one which must be executed in-order, not the memory load uops (so not the whole instruction).

> It is not the case with clang either, see https://godbolt.org/z/fh4q8zc8G

Interesting. I did not test with vector intrinsics. Maybe Clang decides not to optimize this further because it is vectorized manually, or maybe it is just a missed optimization. It works well with scalar. Here is the actual generated code ( see https://godbolt.org/z/j91qj188z ):

```
.LBB0_3:                                # =>This Inner Loop Header: Depth=1
        vaddpd  ymm0, ymm0, ymmword ptr [rax + 8*rdi]
        vaddpd  ymm1, ymm1, ymmword ptr [rax + 8*rdi + 32]
        vaddpd  ymm2, ymm2, ymmword ptr [rax + 8*rdi + 64]
        vaddpd  ymm3, ymm3, ymmword ptr [rax + 8*rdi + 96]
        add     rdi, 16
        cmp     rdx, rdi
        jne     .LBB0_3
```

**Comment by hlide** [comment]
I see rdi is an index and there is 4 vaddpd per iteration, so I would expect a `add rdi, 4` and not `add rdi, 16` or did I miss something?

**Comment by elena901** [comment]
@zephyr111 So Why Clang doesn't perform optimization in the case above? And how we can "cheat" this dependency in case of ISPC (without inventing new `fast-math` mode). Yes i saw the proposed PR, but still want to ask you.

**Comment by zephyr111** [comment]
@elena901 

> So Why Clang doesn't perform optimization in the case above?

Because:

>  compilers *cannot* reorder dependent FP instructions because of the non-associativity of FP numbers defined in the IEEE-754 standard

Re-association is the **only** way Clang can optimize that and this is what FastMath flags enable. It breaks the IEEE-754 standard and can cause significant precision loss (dependent of the actual re-associations done) in pathological cases. In this very-specific case, it can improve slightly the precision. However, the precision is dependent on compiler optimization (so not deterministic between different versions of ISPC using different version of LLVM). Besides, there are cases where fast-math is not an option like a Kahan summation.

At the ISPC level, using a double-pumping target or even quad-pumping helps to mitigate the latency performance issue (my "cheat" what to benefit from a kind of quad-pumping -- or more -- thanks to targets having tiny mask and thus a big `programCount` value). This is another way to perform re-association, but in a way ISPC users fully *control* (deterministic for a given `programCount`). This is critical on some algorithms, like Kahan summation.

**Comment by elena901** [comment]
> my "cheat" what to benefit from a kind of quad-pumping -- or more -- thanks to targets having tiny mask and thus a big `programCount` value.

We need one more dimension for `pipeIndex` and `pipeCount`.

__EDIT:__

Then we need to:

```C
v = reduce_add(v, axis = 1); // xmm0 = xmm1 = ... = xmm_n n - number of workers.
v = reduce_add(v, axis = 0); // all lanes are the same
```

or just __reduce_add(v)__ which is a shortcut for axis=[0,1]. It's just an abstract example, not related to the x86 architecture :slightly_smiling_face:.

**Comment by elena901** [comment]
Probably it was bad naming. Probably it's better:

`workerIndex` and `workerCount`
or
`programIndex[0]` and `programCount[0]`
`programIndex[1]` and `programCount[1]`

So local index will be look like:
```C
#define PROGRAMLINDEX programIndex[1]*programCount[0] + programIndex[0]
```

I'm going to investigate this opportunity deeper.

**Comment by zephyr111** [comment]
@elena901 I do not see how adding a new dimension is supposed to help here. The only parameters impacting the performance of FMAs to consider is the effective instruction throughput, the instruction latency as stated by the Little's law. The effective instruction throughput is impacted by the dependencies/instructions in the code (i.e. the actual ISPC algorithm), the presence of possibly-unnecessary other instructions, and the target CPU (mainly the micro-architecture + frequency). Here, the FMA latency is so big that the additional unnecessary instructions are not actually a problem. This applies on nearly-all modern mainstream CPUs (except very-low-power/in-order CPUs), not even just for x86 (same thing on the Pcore of the Apple's M1 Pro or even apparently POWER9 too). Indeed, the pipeline is wide enough and the latency of FMA is so big compared to other basic instructions in the loop that the (dependent) FMAs are always actually the bottleneck (or close to). This is why unrolling does not help much here. As a result, assuming we cannot control the dependencies (mainly limited by the IEEE-754 rules and dependent of the algorithm and so the ISPC user), nor the target CPU, we can only adjust the number of *independent* FMA instructions in the loop. This is controlled by `programCount` and so the ISPC target. Changing the layout does not help. In the end, this is as simple as this: a higher `programCount` means more FMA in the pipeline and so better performance, unless the pipeline is fully saturated. The Little's law tells us that the maximum (usually close to optimal) number of concurrent FMA is `FMA_per_cycle * FMA_latency` which is typically 8 on modern x86 CPUs (same on the M1 Pro and apparently 10~14 on POWER9). That being said, increasing `programCount` put a significant more pressure on registers (only 16 in AVX2) and other bottlenecks need to be considered too (e.g. memory throughput). In practice, double-pumping often good enough. On top of that, the cost of adding/maintaining a new ISPC target should be considered too.

**Comment by elena901** [comment]
> I do not see how adding a new dimension is supposed to help here.

Yes, not too much. Honestly, it was just a reason for discussion. :slightly_smiling_face:




> Because:
> 
> > compilers _cannot_ reorder dependent FP instructions because of the non-associativity of FP numbers defined in the IEEE-754 standard

My previous question refers to these messages:


> It is not the case with clang either



> Interesting. I did not test with vector intrinsics. Maybe Clang decides not to optimize this further because it is vectorized manually, or maybe it is just a missed optimization.

I haven't checked the result on modern versions yet.

**Comment by elena901** [comment]
> one more dimension

In theory this additional dimension could be hidden from the user and "unroll" automatically according to the context and "reciprocal throughput" on the user machine. [Related](https://github.com/ispc/ispc/issues/1310).

**UPDATE:**

Btw MasPar systems had a C language extension for parallel programming (see [ispc](https://ispc.github.io/papers/ispc_inpar_2012.pdf) paper for the details), where each varying variable could stand for a 2D array/Matrix. I'm looking for this Programming Guide at this moment.
[Related](https://github.com/ispc/ispc/issues/2200)

**Comment by elena901** [comment]
It seems like two different tasks:
1. implement "rolling" a.k.a variable-length axes.
2. and 2D-targets.

**Comment by elena901** [comment]
First of all we need [nested foreach](https://github.com/ispc/ispc/issues/280) loop for these purposes. For this kind of construct:

```ISPC
foreach(y = 0...height, /*axis = 1*/)
   foreach(x = 0...width, /*axis = 0*/)
      array(x,y);
```


**EDIT:**

And uniform for analogue:

```ISPC
for(uniform int y0 = 0; y0 < height; y0+=ProgramCount[1]){
   for(uniform int x0 = 0; x0 < width; x0+=ProgramCount[0]){
      index2D = y0 * width + programIndex[1] * width + x0 + programIndex[0];
   }
}
```

Single foreach loop will stand for:

```ISPC
foreach(i = 0...size, /*axis = {0,1}*/)
```
But i think it would be nice to have an option to specify axes directly.

**EDIT1:**

The core idea behind it is... it's tricky to deal with very long width (let's say x-128) in some cases just in 1D-dimension. 16x8 makes some things much simpler to my point of view.

>implement "rolling" a.k.a variable-length axes.

Additional dimension in the first case is just:

```ISPC
dim1 = programCount / BASE_WIDTH;
```

>put a significant more pressure on registers

>the cost of adding/maintaining a new ISPC target should be considered too

@zephyr111
I understand that the main problems at this moment are: the ability of compiler to deal with much higher "register pressure" and automating generation of "missed targets". When it's possible of course.

**Comment by zephyr111** [comment]
@elena901 

This is a lot to untangle.

I think the proposal of 2D targets clearly needs its own separate space, typically in [discussion](https://github.com/ispc/ispc/discussions). This is not something simple to implement IMHO (it certainly changes significantly the ISPC implementation). More importantly, it changes the language semantics and many points should be clarified/defined. For example, what about the several reduction functions like `reduce_add`, `rotate`, `lanemask`, `extract`, and `exclusive_scan_add`. Also what about the interoperability with C (must not break compatibility). It might require a huge work and I am not convinced it is worth the effort (quad pumping is often enough, or double pumping + SMT-x2 -- see later).

> Btw MasPar systems had a C language extension for parallel programming

MasPar is defunct for (more than) 20 years. I think It made sense on the hardware the company sold but not that much on modern mainstream hardware. Btw, I do not see any direct relation between the posted link and MasPar. 

> implement "rolling" a.k.a variable-length axises

I am not sure what you mean by "variable-length axises". For each target, in many cases, we generate a specialized code so as to improve performance compared to a generic implementation (sometimes generating very inefficient code -- possibly a scalar one). Without a fixed size, we should automatize this process which is (much) more difficult to do. Besides, the interoperability with C (or even SYCL) also needs that.

> For this kind of construct: [...]

Without the axes, the language would be ambiguous to me. I do not expect the compiler to know which axis is targeted. A `foreach` can AFAIK call functions defined in another translation unit itself containing another nested `foreach`. As a result tracking foreach axes seems required (and certainly not trivial). Besides, masking (and its rigorous semantic) should be considered. Btw, please also note that you can already write `foreach(y=0...h, x=0...w) f(y, x);` in ISPC.

How this would apply to a basic sum computation? It seems not much simpler than doing the work manually (i.e. with a `for` + `foreach` + some tricks + manual reductions).

> the ability of compiler to deal with much higher "register pressure"

Note that the problem is not much specific to ISPC. The same can be true in C while vectorizing code manually. The compiler can mitigate this issue a bit but this is not easy to do. I agree with mmp in the provided link that it is certainly good *"for the compiler to not try to be too smart about this sort of thing"* hence providing basic double-pumping target for that. But yes, the quad-pumping targets (or more) are kind of likely to result in register-pressure issues, at least in SSE/AVX, and so not much more interesting than double-pumping.

> automating generation of "missing targets"

Yes, kind of. You should check how it is implemented to better understand the problem. I tried to implement this target 1-2 years ago but AFAIK this changed quite a lot since.

Implementing this target seems quite straightforward to me (while 2D target does not seem to me). I did it some time ago (working prototype supporting like 95% of cases). Maintaining it is a little costly, though. It should be AFAIK better now than when the issue was created. 
By the way, you pointed out the benefit of SMT regarding POWER (apparently in a now-deleted message), which applies on x86: SMT should help to mitigate this problem a bit on parallel code. I missed this because my CPU does not support it (i5-9600KF), unlike ~90% of users of mainstream x86 CPUs (so mine is not representative). It is still a good idea to be able to get good performance in sequential though. Not all codes can be (efficiently) multi-threaded.
In the end, an avx2-i64x8 target is not perfect, but good enough IMHO in most cases. It can be compensated with SMT or some ILP (in the loop) regarding the target use-case. In AVX-512 and AVX-10, which are becoming more widespread, quad-pumping is easy to do in a clean (and often rather-efficient) way, and so rather good enough. When this is not enough, a manual optimization can be performed.

**Comment by elena901** [comment]
> I understand that the main problems at this moment are: the ability of compiler to deal with much higher "register pressure" and automating generation of "missed targets". When it's possible of course.

Not a sprinter distance in other words. And it's still my experimental project at this moment.

>Without the axes

@zephyr111 Yes, of course "axes". Sorry for my T9 🙂 .

You probably don't get the idea. All varying variables are still old good plain data. NxN is just a consideration primarily for matmul and transpose operations.

>A foreach can AFAIK call functions defined in another translation unit itself containing another nested foreach.

Honestly, i don't see any major issues here. Anyway [nested foreach](https://github.com/ispc/ispc/issues/280) loop at this moment is not supported.

>As a result tracking foreach axes seems required (and certainly not trivial).

In case of:

```ISPC
foreach(i = 0...height, /*axis = 1*/)
   another_module( i );
```
varying variable will look like:
(0,0,0,0,1,1,1,1,2,2,2,2,3,3,3,3) for 4x4 targets
and
(0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1) for 2x8 targets.

You don't need any "traction control" here.

**ADDED:**

```ISPC
another_module:

   foreach(j = 0...width, /*axis = 0*/)
      k = j*i;
```

k:
(0,0,0,0,0,1,2,3,0,2,4,6,0,3,6,9) for 4x4
(0,0,0,0,0,0,0,0,0,1,2,3,4,5,6,7) for 2x8

It's the result at the first iteration.

```ISPC
transpose(k);
```

transpose(k):
(0,0,0,0,0,1,2,3,0,2,4,6,0,3,6,9) 4x4

**Comment by elena901** [comment]
> MasPar is defunct for (more than) 20 years. I think It made sense on the hardware the company sold but not that much on modern mainstream hardware. Btw, I do not see any direct relation between the posted link and MasPar.

MasPar MP-1 and MP-2 supercomputers... were programmed in the extended-C language MPL. ... MPL provided vector control constructs with syntax similar to ISPC. MPL provided generalized SIMD pointers similar to the ones in ispc.

**ispc paper, Chapter "Historical systems".**

**Comment by elena901** [comment]
> MasPar is defunct for (more than) 20 years. I think It made sense on the hardware the company sold but not that much on modern mainstream hardware. Btw, I do not see any direct relation between the posted link and MasPar.

C* uses `poly`, FLAP C uses `parallel`, MPL uses `plural`.

The MasPar MP-1 and MP-2 supercomputers ... were programmed in the extended-C language MPL... MPL provided vector control constructs with syntax similar to ISPC. MPL provided generalized SIMD pointers similar to the ones in ISPC.

**ispc paper, Chapters "Uniform Datatypes" and "Historical systems".**

RenderMan pioneered the concept of `uniform` and `varying` variables. `Uniform` variables are only set ones for each set of properties, `varying` variables on the other hand are allowed to change as a function of position. Another data-parallel language is Cn. Cn extends the type system by two modifiers `mono` and `poly`.

**Ingo Wald et al, the IVL paper.**
